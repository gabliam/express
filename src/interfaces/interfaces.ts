import * as express from 'express';
import { interfaces } from '@gabliam/core';


export type HandlerDecorator = (target: any, key: string, value: any) => void;

export interface ExpressPluginConfig {
    rootPath: string;

    port: number;

    hostname: string;
}

export type MiddlewareConfigurator = (...values: any[]) =>  express.RequestHandler |Â express.RequestHandler[];

export interface MiddlewareDefinition {
    name: string;

    values: any[];
}

/**
 * Config function
 */
export type ConfigFunction = (app: express.Application) => void;

export interface ExpressConfigRegistry extends interfaces.ValueRegistry {
    key: string;
}
