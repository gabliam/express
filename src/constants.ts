/** Type in registry */
export const TYPE = {
  Controller: 'ControllerType'
};

export const METADATA_KEY = {
  controller: '_controller',
  controllerMethod: '_controller-method',
  controllerParameter: '_controller-parameter',
  middleware: '_middleware',
  MiddlewareConfig: '_middlewareConfig',
  MiddlewareErrorConfig: '_middlewareErrorConfig'
};

export enum PARAMETER_TYPE {
  REQUEST,
  RESPONSE,
  PARAMS,
  QUERY,
  BODY,
  HEADERS,
  COOKIES,
  NEXT
}

export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const EXPRESS_PLUGIN_CONFIG = Symbol('GABLIAM/EXPRESS_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/EXPRESS_APP');
export const SERVER = Symbol('GABLIAM/EXPRESS_SERVER');
