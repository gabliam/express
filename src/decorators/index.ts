export * from './controller';
export * from './middleware';
export * from './config';
export * from './method';
export * from './params';
