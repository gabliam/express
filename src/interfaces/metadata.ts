import * as express from 'express';
import { MiddlewareDefinition } from './interfaces';
import { inversifyInterfaces } from '@gabliam/core';

export interface ControllerMetadata {
  path: string;

  target: any;

  json: boolean;
}

export interface ControllerMethodMetadata {
  path: string;

  target: any;

  method: string;

  key: string;
}

export interface ExpressConfigMetadata {
  id: inversifyInterfaces.ServiceIdentifier<any>;

  key: string;
}

export type MiddlewareMetadata = express.RequestHandler | MiddlewareDefinition;
