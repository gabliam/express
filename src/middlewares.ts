import { ExpressConfig } from './interfaces';

export class MiddlewareConfig {
  private _middlewares: ExpressConfig[] = [];

  private _errorMiddlewares: ExpressConfig[] = [];


  addMiddleware(middleware: ExpressConfig) {
    this._middlewares.push(middleware);
  }

  addErrorMiddleware(middleware: ExpressConfig) {
    this._errorMiddlewares.push(middleware);
  }

  get middlewares() {
    return this._middlewares.slice();
  }

  get errorMiddlewares() {
    return this._errorMiddlewares.slice();
  }
}
