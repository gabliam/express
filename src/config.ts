import { PluginConfig, Bean, Value } from '@gabliam/core';
import { DEFAULT_ROUTING_ROOT_PATH, EXPRESS_PLUGIN_CONFIG } from './constants';
import { ExpressPluginConfig } from './interfaces';
import * as Joi from 'joi';

@PluginConfig()
export class RestPluginConfig {
    @Value('application.express.rootPath', Joi.string())
    rootPath = DEFAULT_ROUTING_ROOT_PATH;

    @Value('application.express.port', Joi.number().positive())
    port: number = process.env.PORT || 3000;

    @Value('application.express.hostname', Joi.string())
    hostname: string;


    @Bean(EXPRESS_PLUGIN_CONFIG)
    restConfig(): ExpressPluginConfig {
        return {
            rootPath: this.rootPath,
            port: this.port,
            hostname: this.hostname
        };
    }
}
