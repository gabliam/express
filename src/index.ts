import { interfaces as coreInterfaces, inversifyInterfaces, Scan, Registry } from '@gabliam/core';
import { TYPE, METADATA_KEY, EXPRESS_PLUGIN_CONFIG, APP, SERVER } from './constants';
import { getMiddlewares } from './metadata';
import { cleanPath } from './utils';
import * as interfaces from './interfaces';
import * as express from 'express';
import * as d from 'debug';
import * as http from 'http';

const debug = d('Gabliam:Plugin:ExpressPlugin');

export * from './decorators';
export { interfaces, APP, SERVER };

@Scan(__dirname)
export default class ExpressPlugin implements coreInterfaces.GabliamPlugin {
  middlewares: interfaces.ExpressConfig[] = [];

  errorMiddlewares: interfaces.ExpressConfig[] = [];

  bind(container: inversifyInterfaces.Container, registry: Registry) {
    container.bind(APP).toConstantValue(express());
    registry.get(TYPE.Controller)
      .forEach(({ id, target }) => container.bind<any>(id).to(target).inSingletonScope());
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    this.buildExpressConfig(container, registry);
    this.buildControllers(container, registry);
    this.buildExpressErrorConfig(container, registry);
  }

  config(container: inversifyInterfaces.Container, registry: Registry, confInstance: any) {
    if (Reflect.hasMetadata(METADATA_KEY.MiddlewareConfig, confInstance.constructor)) {
      const metadataList: interfaces.ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        this.middlewares.push({ order, instance: confInstance[key].bind(confInstance[key]) })
      });
    }

    if (Reflect.hasMetadata(METADATA_KEY.MiddlewareErrorConfig, confInstance.constructor)) {
      const metadataList: interfaces.ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        this.errorMiddlewares.push({ order, instance: confInstance[key].bind(confInstance[key]) });
      });
    }
  }

  private buildExpressConfig(container: inversifyInterfaces.Container, registry: Registry) {
    const app = container.get<express.Application>(APP);
    this.middlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildExpressErrorConfig(container: inversifyInterfaces.Container, registry: Registry) {
    const app = container.get<express.Application>(APP);
    this.errorMiddlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildControllers(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<interfaces.ExpressPluginConfig>(EXPRESS_PLUGIN_CONFIG);

    debug('registerControllers', TYPE.Controller);
    const controllerIds = registry.get(TYPE.Controller);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: interfaces.ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      const controllerMiddlewares = getMiddlewares(container, controller.constructor);

      const methodMetadatas: interfaces.ControllerMethodMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        controller.constructor
      );

      if (controllerMetadata && methodMetadatas) {
        const router = express.Router();
        const routerPath = cleanPath(`${restConfig.rootPath}${controllerMetadata.path}`);

        debug(`New route : "${routerPath}"`);

        methodMetadatas.forEach((methodMetadata: interfaces.ControllerMethodMetadata) => {
          const methodMetadataPath = cleanPath(methodMetadata.path);
          const methodMiddlewares = getMiddlewares(container, controller.constructor, methodMetadata.key);
          debug(methodMetadataPath);
          debug({ methodMiddlewares, controllerMiddlewares });
          const handler: express.RequestHandler = this.handlerFactory(
            container,
            controllerId,
            methodMetadata.key,
            controllerMetadata.json
          );
          (router as any)[methodMetadata.method](
            methodMetadataPath,
            ...controllerMiddlewares,
            ...methodMiddlewares,
            handler
          );
        });
        const app = container.get<express.Application>(APP);
        app.use(routerPath, router);
      }
    });
  }

  private handlerFactory(
    container: inversifyInterfaces.Container,
    controllerId: any,
    key: string,
    json: boolean
  ): express.RequestHandler {
    return (req: express.Request, res: express.Response, next: express.NextFunction) => {
      const result: any = container.get<any>(controllerId)[key](req, res, next);

      // try to resolve promise
      if (result && result instanceof Promise) {

        result.then((value: any) => {
          if (value !== undefined && !res.headersSent) {
            if (json) {
              res.json(value);
            } else {
              res.send(value);
            }
          }
        })
          .catch((error: any) => {
            next(error);
          });

      } else if (result !== undefined && !res.headersSent) {
        if (json) {
          res.json(result);
        } else {
          res.send(result);
        }
      }
    };
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: inversifyInterfaces.Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>((resolve) => {
        server.close(() => resolve());
      });
    } catch (e) {
    }
  }

  async start(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<interfaces.ExpressPluginConfig>(EXPRESS_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = (typeof port === 'string') ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    function onListening(): void {
      const addr = server.address();
      const bind = (typeof addr === 'string') ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }
}
