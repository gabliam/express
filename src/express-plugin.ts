import {
  interfaces as coreInterfaces,
  inversifyInterfaces,
  Scan,
  Registry
} from '@gabliam/core';
import {
  TYPE,
  METADATA_KEY,
  EXPRESS_PLUGIN_CONFIG,
  APP,
  SERVER,
  PARAMETER_TYPE
} from './constants';
import { getMiddlewares } from './metadata';
import { cleanPath } from './utils';
import {
  ExpressConfigMetadata,
  ExpressPluginConfig,
  ControllerMetadata,
  ControllerMethodMetadata,
  ParameterMetadata,
  ControllerParameterMetadata
} from './interfaces';
import * as express from 'express';
import * as d from 'debug';
import * as http from 'http';
import { MiddlewareConfig } from './middlewares';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Scan(__dirname)
export class ExpressPlugin implements coreInterfaces.GabliamPlugin {
  /**
   * binding phase
   *
   * Bind all controller and bind express app
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    container.bind(APP).toConstantValue(express());
    registry
      .get(TYPE.Controller)
      .forEach(({ id, target }) =>
        container.bind<any>(id).to(target).inSingletonScope()
      );

    container.bind(MiddlewareConfig).toConstantValue(new MiddlewareConfig());
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    this.buildExpressConfig(container, registry);
    this.buildControllers(container, registry);
    this.buildExpressErrorConfig(container, registry);
  }

  /**
   * Management of @middleware decorator in config class
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   * @param  {any} confInstance
   */
  config(
    container: inversifyInterfaces.Container,
    registry: Registry,
    confInstance: any
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    // if config class has a @middleware decorator, add in this.middlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }

    // if config class has a @middleware decorator, add in this.errorMiddlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addErrorMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }
  }

  /**
   * Build express middleware
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  private buildExpressConfig(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    const app = container.get<express.Application>(APP);
    middlewareConfig.middlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  /**
   * Build express error middleware
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  private buildExpressErrorConfig(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    const app = container.get<express.Application>(APP);
    middlewareConfig.errorMiddlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildControllers(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );

    debug('registerControllers', TYPE.Controller);
    const controllerIds = registry.get(TYPE.Controller);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      const controllerMiddlewares = getMiddlewares(
        container,
        controller.constructor
      );

      const methodMetadatas: ControllerMethodMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        controller.constructor
      );

      const parameterMetadata: ControllerParameterMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controllerParameter,
        controller.constructor
      );

      if (controllerMetadata && methodMetadatas) {
        const router = express.Router();
        const routerPath = cleanPath(
          `${restConfig.rootPath}${controllerMetadata.path}`
        );

        debug(`New route : "${routerPath}"`);

        methodMetadatas.forEach((methodMetadata: ControllerMethodMetadata) => {
          let paramList: ParameterMetadata[] = [];
          if (parameterMetadata) {
            paramList = parameterMetadata[methodMetadata.key] || [];
          }
          const methodMetadataPath = cleanPath(methodMetadata.path);
          const methodMiddlewares = getMiddlewares(
            container,
            controller.constructor,
            methodMetadata.key
          );
          debug(methodMetadataPath);
          debug({ methodMiddlewares, controllerMiddlewares });
          const handler: express.RequestHandler = this.handlerFactory(
            container,
            controllerId,
            methodMetadata.key,
            controllerMetadata.json,
            paramList
          );
          (router as any)[methodMetadata.method](
            methodMetadataPath,
            ...controllerMiddlewares,
            ...methodMiddlewares,
            handler
          );
        });
        const app = container.get<express.Application>(APP);
        app.use(routerPath, router);
      }
    });
  }

  private handlerFactory(
    container: inversifyInterfaces.Container,
    controllerId: any,
    key: string,
    json: boolean,
    parameterMetadata: ParameterMetadata[]
  ): express.RequestHandler {
    return (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const controller = container.get<any>(controllerId);
      const args = this.extractParameters(
        controller,
        key,
        req,
        res,
        next,
        parameterMetadata
      );
      const result: any = controller[key](...args);

      // try to resolve promise
      if (result && result instanceof Promise) {
        result
          .then((value: any) => {
            if (value !== undefined && !res.headersSent) {
              if (json) {
                res.json(value);
              } else {
                res.send(value);
              }
            }
          })
          .catch((error: any) => {
            next(error);
          });
      } else if (result !== undefined && !res.headersSent) {
        if (json) {
          res.json(result);
        } else {
          res.send(result);
        }
      }
    };
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: inversifyInterfaces.Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    function onListening(): void {
      const addr = server.address();
      const bind =
        typeof addr === 'string' ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }

  private extractParameters(
    target: any,
    key: string,
    req: express.Request,
    res: express.Response,
    next: express.NextFunction,
    params: ParameterMetadata[]
  ): any[] {
    const args = [];
    if (!params || !params.length) {
      return [req, res, next];
    }
    const getParam = this.getFuncParam(target, key);

    for (const item of params) {
      switch (item.type) {
        default:
          args[item.index] = res;
          break; // response
        case PARAMETER_TYPE.REQUEST:
          args[item.index] = getParam(req, null, item);
          break;
        case PARAMETER_TYPE.NEXT:
          args[item.index] = next;
          break;
        case PARAMETER_TYPE.PARAMS:
          args[item.index] = getParam(req, 'params', item);
          break;
        case PARAMETER_TYPE.QUERY:
          args[item.index] = getParam(req, 'query', item);
          break;
        case PARAMETER_TYPE.BODY:
          args[item.index] = getParam(req, 'body', item);
          break;
        case PARAMETER_TYPE.HEADERS:
          args[item.index] = getParam(req, 'headers', item);
          break;
        case PARAMETER_TYPE.COOKIES:
          args[item.index] = getParam(req, 'cookies', item);
          break;
      }
    }
    args.push(req, res, next);
    return args;
  }

  private getFuncParam(target: any, key: string) {
    return (
      source: any,
      paramType: string | null,
      itemParam: ParameterMetadata
    ) => {
      const name = itemParam.parameterName;

      let param = source;
      if (paramType !== null && source[paramType]) {
        param = source[paramType];
      }

      let res = param[name];
      if (res) {
        /**
         * For query, all value sare considered to string value.
         * If the query waits for a Number, we try to convert the value
         */
        if (paramType === 'query' || paramType === 'params') {
          const type: Function[] = Reflect.getMetadata(
            'design:paramtypes',
            target,
            key
          );
          if (Array.isArray(type) && type[itemParam.index]) {
            try {
              if (type[itemParam.index].name === 'Number') {
                // parseFloat for compatibility with integer and float
                res = Number.parseFloat(res);
              }
            } catch (e) {}
          }
        }
        return res;
      } else {
        return paramType === 'query' ? undefined : param;
      }
    };
  }
}
