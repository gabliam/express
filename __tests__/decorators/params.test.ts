import {
  Request,
  Response,
  RequestParam,
  QueryParam,
  RequestBody,
  RequestHeaders,
  Cookies,
  Next,
  Get,
  Post
} from '../../src/decorators';

import { ControllerParameterMetadata } from '../../src/interfaces';
import { METADATA_KEY } from '../../src/constants';

describe('params decorators', () => {
  test('should add parameter metadata to a class when decorated with @params', () => {
    const middleware = [
      function() {
        return;
      }
    ];

    class TestController {
      @Get('/foo/:id', ...middleware)
      public test(
        @RequestParam('id') id: any,
        @QueryParam('cat') cat: any,
        @Request() req: any,
        @Response() res: any
      ) {
        return;
      }

      @Post('foo')
      public test2(@RequestBody('dog') dog: any) {
        return;
      }

      @Get('lol')
      public test3(
        @RequestHeaders('Auth') auth: any,
        @Cookies() cookies: any,
        @Next() next: any
      ) {
        return;
      }
    }
    const methodMetadataList: ControllerParameterMetadata = Reflect.getMetadata(
      METADATA_KEY.controllerParameter,
      TestController
    );
    expect(methodMetadataList).toMatchSnapshot();
  });
});
