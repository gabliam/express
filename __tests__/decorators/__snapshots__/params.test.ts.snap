// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`params decorators should add parameter metadata to a class when decorated with @params 1`] = `
Object {
  "test": Array [
    Object {
      "index": 0,
      "parameterName": "id",
      "type": 2,
    },
    Object {
      "index": 1,
      "parameterName": "cat",
      "type": 3,
    },
    Object {
      "index": 2,
      "parameterName": "default",
      "type": 0,
    },
    Object {
      "index": 3,
      "parameterName": "default",
      "type": 1,
    },
  ],
  "test2": Array [
    Object {
      "index": 0,
      "parameterName": "dog",
      "type": 4,
    },
  ],
  "test3": Array [
    Object {
      "index": 0,
      "parameterName": "Auth",
      "type": 5,
    },
    Object {
      "index": 1,
      "parameterName": "default",
      "type": 6,
    },
    Object {
      "index": 2,
      "parameterName": "default",
      "type": 7,
    },
  ],
}
`;
